### PR TITLE
Recursively return all the generated files

### PR DESCRIPTION
**Pull request by keynmol**
Use `FileFunction` mechanism to more accurately cache the input and output files of the task, as well as react to file changes for more interactive experience.

This combination _seems_ to remove the issues I was seeing before with Zinc.

### CODE REVIEW

**Comment by keynmol** [comment]
I think this needs a bit more work, as I'm getting 

```
[error] 6 |import smithy4s.syntax._
[error]   |       ^^^^^^^^^^^^^^^
[error]   |       value syntax is not a member of smithy4s
```

occasionally when modifying a broken compilation unit..

File: modules/codegen-plugin/src/smithy4s/codegen/CodegenPlugin.scala [hunk 1]
```diff
@@ -172,6 +176,16 @@ object Smithy4sCodegenPlugin extends AutoPlugin {
         .toList
 
       resPaths.map(path => new File(path.toString))
-    } else (conf / smithy4sOutputDir).value.listFiles().toSeq
+    } else {
+      val base = (conf / smithy4sOutputDir).value
+
+      val allOutputFiles = (base ** "*").get()
```
Comment (Baccata) [inline modules/codegen-plugin/src/smithy4s/codegen/CodegenPlugin.scala:182]: I'm not certain that the logic is correct, considering smithy4s doesn't wipe the output folder between runs. Can't we use a similar mechanism to the one you used, but for outputs ?

**Comment by Baccata** [comment]
@keynmol is this ready ?

**Comment by keynmol** [comment]
I believe so, yes.
It's better than what's currently in main, so that's 1 reason to merge it :)

**Review by Baccata — APPROVED** [review]

**Pull request merged**